### DISCUSSION TITLE
Form and table id not appended to table cells

### DISCUSSION THREAD

**Discussion by promodsingh** [primefaces/community#3162] [PrimeFaces] [answered]
Hello, 

I am creating datatable and it's column programmatically. But when the table and corresponding columns rendered on UI form and table id is not appended to the component inside cell or even headers. Id's are just like j_id7 instead of form:<Table_Name>:<id>_head.

Is there any example on how to programmatically create columns and table or if you can point out the mistake.

Sample code for table and column creation

`
Datatable table = new datatable()
table.setSelectionMode( "single" );
table.setVar( "address" );
table.setWidgetVar( "addressTable" );
table.setValueExpression( "rowKey", "address", "id", String.class );
table.setRowIndexVar( "rowIndex" );

for ( ColumnsModel col : columnsList ) {
	Column column = new Column();
	column.setHeaderText( col.getLabel() );
	column.setWidth( String.valueOf( col.getWidth() ) );
	column.setSortable( true );
	uiColumns.add( column );
}

table.setColumns( uiColumns );
`

Thanks

**Comment by melloware** [comment]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

**Comment by tandraschko** [comment, marked as answer]
just 2 hints from my side:
1) never use the constructor of a component, JSF have a #createComponent API to use since ages
2) table.setColumns isnt the right way. You need to attach it as child component. the Datatable will then later build the columns list itself.

anyway, the way you trying to create your view isnt broadly used and therefor not much people can you help you here. i would make it via xhtml.

**Reply by melloware** [reply]
Agree with @tandraschko if you are creating a view or components dynamically here is a working example how to do it.

```java
package org.primefaces.test;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.faces.annotation.View;
import jakarta.faces.application.StateManager;
import jakarta.faces.component.UIComponent;
import jakarta.faces.component.UIOutput;
import jakarta.faces.component.html.HtmlBody;
import jakarta.faces.component.html.HtmlCommandButton;
import jakarta.faces.component.html.HtmlDoctype;
import jakarta.faces.component.html.HtmlForm;
import jakarta.faces.component.html.HtmlOutputText;
import jakarta.faces.context.FacesContext;
import jakarta.faces.view.facelets.Facelet;

@View("/facelet.xhtml")
@ApplicationScoped
public class FaceletView extends Facelet {

    @Override
    public void apply(FacesContext facesContext, UIComponent parent) {
        if (!facesContext.getAttributes().containsKey(StateManager.IS_BUILDING_INITIAL_STATE)) {
            return;
        }

        var components = new ComponentBuilder(facesContext);
        var rootChildren = parent.getChildren();

        var htmlDoctype = new HtmlDoctype();
        htmlDoctype.setRootElement("html");
        rootChildren.add(htmlDoctype);

        UIOutput output = new UIOutput();
        output.setValue("<html xmlns=\"http://www.w3.org/1999/xhtml\">");
        output.getAttributes().put("escape", false);
        rootChildren.add(output);

        HtmlBody body = components.create(HtmlBody.COMPONENT_TYPE);
        rootChildren.add(body);

        HtmlForm form = components.create(HtmlForm.COMPONENT_TYPE);
        form.setId("form");
        body.getChildren().add(form);

        HtmlOutputText message = components.create(HtmlOutputText.COMPONENT_TYPE);
        message.setId("message");

        HtmlCommandButton actionButton = components.create(HtmlCommandButton.COMPONENT_TYPE);
        actionButton.setId("button");
        actionButton.addActionListener(
                e -> message.setValue("Hello, World! Welcome to Faces 4.0 on Jakarta EE 10"));
        actionButton.setValue("Greet");

        form.getChildren().add(actionButton);

        body.getChildren().add(message);

        output = new UIOutput();
        output.setValue("</html>");
        output.getAttributes().put("escape", false);
        rootChildren.add(output);
    }

    private static class ComponentBuilder {
        FacesContext facesContext;

        ComponentBuilder(FacesContext facesContext) {
            this.facesContext = facesContext;
        }

        @SuppressWarnings("unchecked")
        <T> T create(String componentType) {
            try {
                return (T) facesContext.getApplication().createComponent(componentType);
            } catch (ClassCastException e) {
                throw new IllegalArgumentException("Component type " + componentType + " is not valid.", e);
            }
        }
    }
}
```

**Comment by promodsingh** [comment]
Thanks @tandraschko,

The issue was the way I was adding the columns. Table.getChildren().add( Column ) solved the issue.
